cpp/ql/lib/semmle/code/cpp/security/BufferWrite.qll

```diff
@@ -69,15 +69,30 @@ abstract class BufferWrite extends Expr {
    * Gets an upper bound to the amount of data that's being written (if one
    * can be found).
    */
-  int getMaxData() { none() }
+  int getMaxData() { result = max(getMaxData(_)) }
```

Comment:
Is this change really backward compatible? Previously, if a customer has written their own class that extends `BufferWrite` and implemented `getMaxData()` like so:
```ql
class MyWrite extends BufferWrite {
  MyWrite() { ... }

  override int getMaxData() { ... }
}
```
then our existing queries correctly pick up these classes and use their `getMaxData()` implementations.

But after this PR, they now override a `getMaxData()` that already has an implementation, but they now have a class where `getMaxData(BufferWriteEstimationReason reason)` implemented as `none()`. And so our queries (that now use the `getMaxData(BufferWriteEstimationReason reason)` version) won't pick up their class.

Does that make sense? Any idea how we can avoid this issue?

Reply:
would it be too rude to mark `getMaxData` as `final`, at leat to not silently not working?

Reply:
I hope we can avoid breaking customer's code like that, but that's at least better than silently removing alerts🤞. In any case, we should probably mark `getMaxData` as `deprecated`.

How about we implement a default version of `getMaxData(BufferWriteEstimationReason reason)` in terms of `getMaxData()` and either:
- Default the reason as `TypeBoundsAnalysis`, or
- Add a dummy reason IPA branch for this case?

Reply:
Hm, or maybe this will actually create non-monotonic recursion? 🤔. Let's see what happens!

Reply:
Ok, I added a `NoSpecifiedReason` case used only for this. The current state of the PR will generate some warnings during library compilation, but no warnings in existing queries as far as I can tell from the tests.

Reply:
Sounds good. Is the warning because we now override a `deprecated` predicate?

Reply:
yes, in order to make a client class that override `getMaxData\0` work with our queries that use `getMaxData\1` now, the default implementation of `getMaxData\1` must call deprecated `getMaxData\0`. So when compiling the library a warning is produced. However the warnings are not produced when compiling/running the queries themselves.

Reply:
Yeah, that makes sense. I guess that's okay.

Reply:
hmm, actually it turns out it's not ok for CI to have warnings in the libraries :frowning_face: Any idea what to do @MathiasVP ? It seems there is no way we can deprecate a predicate that could be overridden by clients and both make those overrides still work as expected and not use deprecated predicates in our own library code without a warning... Are there some secret pragmas to suppress warnings maybe?

Reply:
As far as I know, there's no such secret pragma (but maybe there should be). 🤔

Reply:
One easy solution is to _not_ use the deprecated keyword for the old `getMaxData` predicate.